modules/core/04-channel/keeper/upgrade.go

```diff
@@ -57,43 +61,229 @@ func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID strin
+	}
+
+	// make sure connection is OPEN
+	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
+		return types.Upgrade{}, errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	// check the connection associated with the passed in connection hops by the chain sending the TryUpgrade
+	// matches the connection associated with the counterparty proposed upgrade
+
+	connection, err := k.GetConnection(ctx, proposedConnectionHops[0])
+	if err != nil {
+		return types.Upgrade{}, err
+	}
+
+	counterpartyHops := counterpartyProposedUpgrade.Fields.ConnectionHops
+	if counterpartyHops[0] != connection.GetCounterparty().GetConnectionID() {
+		return types.Upgrade{}, errorsmod.Wrapf(connectiontypes.ErrInvalidConnection, "proposed connection hops (%s) does not match counterparty proposed connection hops (%s)", proposedConnectionHops, counterpartyHops)
+	}
+
+	// construct counterpartyChannel from existing information and provided counterpartyUpgradeSequence
+	counterpartyChannel := types.Channel{
+		State:           types.INITUPGRADE,
+		Counterparty:    types.NewCounterparty(portID, channelID),
+		Ordering:        channel.Ordering,
+		ConnectionHops:  counterpartyProposedUpgrade.Fields.ConnectionHops,
+		Version:         channel.Version,
+		UpgradeSequence: counterpartyUpgradeSequence,
+	}
+
+	// crete upgrade fields from counterparty proposed upgrade and own verified connection hops
+	upgradeFields := types.NewUpgradeFields(
+		counterpartyProposedUpgrade.Fields.Ordering,
+		proposedConnectionHops,
+		counterpartyProposedUpgrade.Fields.Version,
+	)
+
+	// if OPEN, then initialize handshake with upgradeFields
+	// this should validate the upgrade fields, set the upgrade path and set the final correct sequence.
+	var proposedUpgrade types.Upgrade
+	if channel.State == types.OPEN {
+
+		if err := k.ValidateUpgradeFields(ctx, upgradeFields, channel); err != nil {
+			return types.Upgrade{}, err
+		}
+
+		proposedUpgrade, err = k.constructProposedUpgrade(ctx, portID, channelID, upgradeFields, proposedUpgradeTimeout)
+		if err != nil {
+			return types.Upgrade{}, errorsmod.Wrap(err, "failed to construct proposed upgrade")
+		}
+
+		// upgrade is blocked on this channelEnd from progressing until flush completes on both ends
+		// packet flushing logic should be performed here, as well as proof verification for the upgrade sequence
+		// and the counterparty channel state
+		if err := k.StartFlushUpgradeHandshake(
+			ctx,
+			portID,
+			channelID,
+			proposedUpgrade.Fields,
+			counterpartyChannel,
+			counterpartyProposedUpgrade,
+			counterpartyUpgradeSequence,
+			types.TRYUPGRADE,
+			proofCounterpartyChannel,
+			proofUpgrade,
+			proofHeight,
+		); err != nil {
+			return types.Upgrade{}, errorsmod.Wrap(err, "failed to start flush upgrade handshake")
+		}
+
```

Comment:
this seems like odd placement. I think we still need to nail down verification in relation to checks which need to use verified fields. Is it possible to reduce the scope of this pr so we can merge some of these changes? Maybe we could leave everything after the above checks as a future todo?

Reply:
Indeed, if this is called in both channel states then it doesn't need to be within the scope of this conditional.

Reply:
yea, it's just bc the proposedUpgrade changes depending on if it's a current Upgrade in crosisng hellos or a new one submitted in the params